Adding Ridgeplot to Arviz-Plots

imperorrp opened the pull request:
First commit for adding ridge plot. As per issue #7, the already implemented forest plot structure was taken as base logic for this plot and the conventions for adding a new plot (https://arviz-plots.readthedocs.io/en/latest/contributing/new_plot.html) have been followed. This plot plots kde ridges instead of credible intervals and point estimates. 

I also added two top-level arguments to this plot since I thought they might be useful- 

- `rescale_kde` (rescales kde by raising the 'y' coord values in the density dataset to a power, set to 1 by default. Using exponent here to amplify the peaks so they're more prominently visible because there's a tendency for kdes to appear very flat when plot_ridge() has a lot of subplots stacked)

- `plot_ridge_base` (boolean to choose whether or not to plot a baseline for each ridge. I think this helps in legibility if the ridges overlap each other or have high peaks, especially if the kde was rescaled)

Also modified the `line_xy` visual element function to accept an additional `y` argument (set to None by default) which is added to `da.sel(plot_axis='y')` if it is not None, so that each ridge is plotted at the correct height corresponding to its labels on the left column. Since the functionality is the same as previously if `y` is not provided, this ensures the function works where used already (like in `plot_dist()`) without deprecation. 

<!-- readthedocs-preview arviz-plots start -->
----
📚 Documentation preview 📚: https://arviz-plots--57.org.readthedocs.build/en/57/

<!-- readthedocs-preview arviz-plots end -->

imperorrp commented:
Some examples of the plot output with the `centered_eight` data.

## Without the extra top-level args for rescaling/adding a baseline:
`azp.plot_ridge(data)`
![image](https://github.com/arviz-devs/arviz-plots/assets/24991131/94a2f73f-109f-4efd-bfa2-58f2cb5e21d5)

`azp.plot_ridge(data, var_names=["mu", "tau"])`
![image](https://github.com/arviz-devs/arviz-plots/assets/24991131/1b8d08b8-142d-4b32-b941-ba23d1e9a2da)

---------
## With rescaling:

`azp.plot_ridge(data, rescale_kde=0.5)`
![image](https://github.com/arviz-devs/arviz-plots/assets/24991131/8dc01040-42bf-47e5-acb3-9968aef48074)

`azp.plot_ridge(data, var_names=["mu", "tau"], rescale_kde=0.5)`
![image](https://github.com/arviz-devs/arviz-plots/assets/24991131/84597452-9f81-4404-8c4e-85a4b2d24344)

`azp.plot_ridge(data, rescale_kde=1.5)`
![image](https://github.com/arviz-devs/arviz-plots/assets/24991131/1abbc381-de14-40d9-832f-2b4cdb0b1338)

`azp.plot_ridge(data, var_names=["mu", "tau"], rescale_kde=1.5)`
![image](https://github.com/arviz-devs/arviz-plots/assets/24991131/5af41183-93b1-47fb-949d-474d1bc28b87)

-------------
## With base line added:

`azp.plot_ridge(data, plot_ridge_base=True)`
![image](https://github.com/arviz-devs/arviz-plots/assets/24991131/91e22268-9ce8-4d74-ac9a-563ec8249bb5)

`azp.plot_ridge(data, var_names=["mu", "tau"], plot_ridge_base=True)`
![image](https://github.com/arviz-devs/arviz-plots/assets/24991131/61e0b221-d3be-41fd-8977-602a04524e68)

(Looks a bit cluttered when baselines are added when there's a lot of ridges but for the latter I think helps in seeing the density, especially if some ridges are overlapping.)

OriolAbril commented on src/arviz_plots/plots/ridgeplot.py:
```diff
@@ -0,0 +1,505 @@
+    default_color = plot_bknd.get_default_aes("color", 1, {})[0]
+    # rescaling kde
+    print(f"\n kde density pre rescaled = {density!r}")
+    for var in density.data_vars:
+        density[var].loc[{"plot_axis": "y"}] = density[var].sel(plot_axis="y") ** rescale_kde
```
This type of transformation should not be an option. kdes are already hard enough to interpret (even if they look simple and we believe we can read them with ease that is not the case). I would not add this as this makes it basically impossible to actually read anything from the kde other than the range and the position of the peak.

By rescaling I meant multiplying all kdes by a common scale factor in order to preserve all proportions in the curve but having the tallest one peak at height 1 (or something else, this could be user modified, similarly to the overlap control argument there is in current `plot_forest`).

```suggestion
    for var in density.data_vars:
        density.loc[{"plot_axis": "y"}] = density.sel(plot_axis="y") / density.sel(plot_axis="y").max().to_array().max() * ridge_height
```

where `ridge_height` regulates the height of the ridge (that is, the height of the tallest peak in the ridge). If `ridge_height=1` (or lower) there should be no overlap at all (when combined=True and there aren't multiple models) with higher values there will be some area overlap (>1 but smallish should mean area overlap but no edge overlap generally, >>1 will generally mean both are and edge overlap). Once both area and edge plotting are working, we should try several values like 1, 1.5, 2 and see with the example arviz datasets how each looks.

OriolAbril commented on src/arviz_plots/visuals/__init__.py:
```diff
@@ -13,13 +13,20 @@
 
-def line_xy(da, target, backend, **kwargs):
+def line_xy(da, target, backend, y=None, **kwargs):
     """Plot a line x vs y.
 
     The input argument `da` is split into x and y using the dimension ``plot_axis``.
+    If an additional y argument is provided, y is added to the values in the `da`
+    dataset sliced along plot_axis='y'.
     """
     plot_backend = import_module(f"arviz_plots.backend.{backend}")
-    return plot_backend.line(da.sel(plot_axis="x"), da.sel(plot_axis="y"), target, **kwargs)
+    da_has_y = "plot_axis" in da.dims and "y" in da.plot_axis
+    if da_has_y:
+        y = da.sel(plot_axis="y") if y is None else da.sel(plot_axis="y") + y
+    if y is None:
+        raise ValueError("Unable to find values for y in `da`")
+    return plot_backend.line(da.sel(plot_axis="x"), y, target, **kwargs)
```
This should use the same pattern as `annotate_label` or `point_estimate_text`. That is

1. Both x and y are arguments of the function, both default to `None`
2. the actual x and y values are obtained with `x, y = _process_da_x_y(da, x, y)` (combining values from the function arguments and the different dataarray coordinate values as needed).

OriolAbril reviewed (commented):
There should be two visual elements that generate the ridge: the edge (already implemented and working) and the face (filling the area of the kde). This will match the behaviour of current plot_ridge (via plot_forest), see for example: https://python.arviz.org/en/stable/examples/plot_forest_ridge.html.

As always either or both can be turned off. Regarding defaults, I think having the edge have the default color is great, for the area I think using the same color as the default would mean having an data-ink ratio too low, we could use `color="white", alpha=0.4` or something like that to help with visualizing overlap between ridges but keeping a "clean" look (that would look quite weird for backgrounds other than white though like `arviz-darkgrid` style), or we could use `plot_kwargs.setdefault("face", False)` to skip it by default but giving users the option to turn it on.

API question (needs discussion before any implementation): I think it would probably be nice to have a shortcut/syntactic sugar to have both edge and face use the same properties, but I am not sure how much worth it it would be.

imperorrp commented:
Renamed the ridge artist to edge and added the face artist. I had to modify the `fill_between_y` visual element function a bit as well so that the 'y' aesthetic argument passed to it by `pc.map` is considered and also updated the `line_xy` visual element with the pattern those other visual element functions use. 

Also moved the kde rescaling code to just after statistical computation so that a new dataset `face_density` can be created and later fed to the `fill_between_y` function. 

About the default color for face- I've just used the same one as for edge for now but with the alpha=0.4 added. Have left the bit about face/edge using the same properties for now as well, for after discussion like you said. 

Also added some tests, with style drawn from the existing ones for `plot_dist` and `plot_forest`. Tried adding some hypothesis tests but there were issues with that so I haven't committed those changes yet. 

Here's how the plots look like now: 

### When combined==True (by default)
![image](https://github.com/arviz-devs/arviz-plots/assets/24991131/dc39c178-66ae-4d74-9ed6-a7091162b38f)

![image](https://github.com/arviz-devs/arviz-plots/assets/24991131/60aa6be9-4bda-4cf8-b45e-19fa1710ee9a)

### When combined==False

![image](https://github.com/arviz-devs/arviz-plots/assets/24991131/a02d27d4-c08a-4cd5-9c5a-a7f589705b78)

![image](https://github.com/arviz-devs/arviz-plots/assets/24991131/5632b7d9-2c4b-47c5-b4ac-e1031646e82c)

OriolAbril reviewed (commented):
> Tried adding some hypothesis tests but there were issues with that so I haven't committed those changes yet.

Try manually generating a plot_ridge without the edge and another without the face. The hypothesis test if you copied it from the plot_forest one is failing at least due to that

OriolAbril commented on src/arviz_plots/visuals/__init__.py:
```diff
@@ -77,8 +84,12 @@ def ecdf_line(values, target, backend, **kwargs):
-def fill_between_y(da, target, backend, *, x=None, y_bottom=None, y_top=None, **kwargs):
+def fill_between_y(da, target, backend, *, x=None, y_bottom=None, y=None, y_top=None, **kwargs):
     """Fill the region between to given y values."""
+    if y_bottom is None and y:
```
Side comment is that `if y_bottom is None and y` would evaluate `y` as a boolean, thus, `y_bottom=None, y=0` would not enter the if. Not relevant with the other behavioural comment though

OriolAbril commented on src/arviz_plots/visuals/__init__.py:
```diff
@@ -77,8 +84,12 @@ def ecdf_line(values, target, backend, **kwargs):
-def fill_between_y(da, target, backend, *, x=None, y_bottom=None, y_top=None, **kwargs):
+def fill_between_y(da, target, backend, *, x=None, y_bottom=None, y=None, y_top=None, **kwargs):
     """Fill the region between to given y values."""
+    if y_bottom is None and y:
+        y_bottom = y
+    if y_top is None and y:
+        y_top = y
```
I think I would modify this to make sure things behave as expected. `x` and `y` are mostly used so we can have aesthetics `x` and `y` that control shifting the visual elements within a plot. Therefore, I would expect that if I pass `y`, everything is shifted along the vertical axes. With these checks, if I give both `y_top` and `y` this won't happen.

I would use

```python
if y is not None:
    y_top += y
    y_bottom += y
```

once the `if "kwarg" in da.dims:` finishes as then both y_top and y_bottom should be array_like elements and not None.

OriolAbril reviewed (commented):
I have added a couple comments for the visuals, to make sure we can combine `x` and `y` as aesthetics along with the data provided by the plots.

imperorrp commented:
> > Tried adding some hypothesis tests but there were issues with that so I haven't committed those changes yet.
> 
> Try manually generating a plot_ridge without the edge and another without the face. The hypothesis test if you copied it from the plot_forest one is failing at least due to that

Yeah I just checked this- since the current code calls the density calculation function only if `edge_kwargs is not False`, it doesn't work when edge is turned off with plot_kwargs. That might be the reason behind the test failing. I could move the density computation outside to fix this, with perhaps a check for if `edge_kwargs or face_kwargs`.

imperorrp commented:
Modified density calculation to occur if either one of edge_kwargs or face_kwargs is not False, indicating that it is required to be calculated for plotting to be possible. Also added the visual element function changes and hypothesis tests code, which appears to pass now for plot_ridge

imperorrp commented:
### Current plot outputs:

#### With both edge and face:
![image](https://github.com/arviz-devs/arviz-plots/assets/24991131/92135139-fc4f-4fb4-804b-66b3aec1e3d3)

#### With edge only: 
![image](https://github.com/arviz-devs/arviz-plots/assets/24991131/7b6dcdec-1420-41a2-8a23-a8d473659ed8)

#### With face only: 
![image](https://github.com/arviz-devs/arviz-plots/assets/24991131/347afffa-a105-41a1-b5bf-f437bb4adfea)

OriolAbril commented on src/arviz_plots/visuals/__init__.py:
```diff
@@ -13,13 +13,20 @@
     """
     plot_backend = import_module(f"arviz_plots.backend.{backend}")
-    return plot_backend.line(da.sel(plot_axis="x"), da.sel(plot_axis="y"), target, **kwargs)
+    if x is not None or y is not None:  # if either x or y is provided as an arg
+        x, y = _process_da_x_y(da, x, y)
+    else:
+        x = da.sel(plot_axis="x")
+        y = da.sel(plot_axis="y")
```
```suggestion
    x, y = _process_da_x_y(da, x, y)
```
